[arkarsg] iP

arkarsg opened the pull request:
# Duke v0.1

> What's on your mind today?

A simple chatbot that manages things you need to do. With all the productivity apps out there, sometimes _all you need_ is an application that is:
- Fuss free
- Easy to use
- Quick

### **Without** all the clutter, focus and get things done 💯.

---

# 🙋‍♂️ Getting Started

****[IMPORTANT]****: Ensure you have Java installed on your device.

1. Download `Duke.jar` from below or click [here](https://github.com/arkarsg/ip/releases/download/A-Jar/Duke.jar)

#### For Mac users
1. Launch your Terminal and navigate to `Duke.jar` file using the `cd` command.
2. Start Duke with `java -jar Duke.jar`

#### For Windows users
1. Launch your Terminal and navigate to `Duke.jar` file.
2. Start Duke with `java -jar Duke.jar`

### Quitting Duke
1. Simply quit with the `bye` command.
---
# 🧐 Features

To help you manage your everyday tasks, here's what the Duke can do for you.
- [x] Note down a simple task to do.
- [x] Note down a task that has a deadline.
- [x] Note down an event.
- [x] Find a task with keywords.
---
# 📖 Documentation

### `todo {task description}`
- Adds a simple to-do task with the given description.
- Accepts Integers and String.

### `deadline {task description} /by {date time}`
- Adds a deadline task with the given `{date time}` as the deadline
- `{date time}` follows ***dd/mm/yyyy hhmm*** (i.e 23/01/2023 1600)

### `event {task description} /from {date time} /to {date time}`
- Adds an event task with the given `{date time}` as the deadline
- `{date time}` follows ***dd/mm/yyyy hhmm*** (i.e 23/01/2023 1600)

### `list`
- Shows an indexed list of all tasks the Duke remembers.

### `delete {task index}`
- Deletes the task associated with the task index from task list.

### `mark {task index}` and `unmark {task index}`
- Mark an unfinished task as completed.
- Unmark removes the completed status from a task.

### `find {search keywords}`
- Returns a list of tasks containing the keywords.

### `bye`
- Exits the application.
- Task list from the current session is saved and loaded the next time Duke is opened.
---
# 🛠 Under the Hood

Pseudocode snippet below shows the main mechanism of how Duke works.

```java
// main loop to accept user inputs
while (isNotBye) {
    handle(input);
}

// Command handler
String[] args = paser.parse(input);
String command = args[0];
commandList.get(command).apply(args);
```

Commit:
Add Gradle support

arkarsg committed:
test

arkarsg committed:
Delete test.txt

arkarsg committed:
Completed Level-1 to echo inputs and quit on command.

arkarsg committed:
Merge commit '8735837f6e5574c95a5cd06b650e16c6447540b7'

* commit '8735837f6e5574c95a5cd06b650e16c6447540b7':
  Delete test.txt

arkarsg committed:
Add ability to store any text as Task object and display as list.

arkarsg committed:
Implemented mark and unmark

arkarsg committed:
Added new Task classes and typesafe ArrayList of tasks

arkarsg committed:
Duke handles ToDo, Deadline and Event. Other strings show "invalid command"

arkarsg committed:
A-TextUITesting. Tasks now cannot be marked/unmarked if they are already in that state.

arkarsg committed:
Fixed Text-ui-test

arkarsg committed:
Error handling with "InvalidCommandException" specific to Duke.

arkarsg committed:
"Delete" implemented

arkarsg committed:
Fixed some formatting issues

arkarsg committed:
Deleted unused variables

arkarsg committed:
Removed unused variables and packages

arkarsg committed:
refactor: command methods are wrapped in HashMap
---------------------------------------------------------------------------------
Commands are now handled with a HashMap storing a string and a consumer. This removes the need for switch-case/if-else blocks and allow easier expansion for growing list of commands in the future.

arkarsg committed:
feat: Added rudimentary save function.

All objects of class Task and its children are now serialisable. save feature serialises all tasks present at the end of the session and writes to duke.txt.

arkarsg committed:
fix: Fixed duke.txt directory to save

arkarsg committed:
fix: Removed save features to support development in its own branch

arkarsg committed:
Revert "fix: Removed save features to support development in its own branch"

This reverts commit c030cf8baecea6d528be5f1a6b5718cada42d3c4.

arkarsg committed:
test branch push

arkarsg committed:
feat: Save/ Load feature implemented.

Duke can now save each session to a .txt file and load from that file in the future sessions. Save takes in the ArrayList of Tasks and serializes through the Serializable interface. The byte stream is then written to 'duke.txt' file. The object is then retrieved in the future session. This is immediately assigned to the new session's ArrayList of Tasks.

If directory to duke.txt does not exist, it is created. If the directory and file exists, it is overwritten to the latest session of save.

arkarsg committed:
fix: removed redundant lines of code.

arkarsg committed:
feat: Events and Deadlines now converts string to DateTime object.

This feature supports different formats for input and display to users for readability

arkarsg committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  feat: Events and Deadlines now converts string to DateTime object.

arkarsg committed:
Add DukeIO

arkarsg committed:
Fix UI messages and error messages

arkarsg committed:
fix: DukeIO implementation.

arkarsg committed:
Add A-MoreOOP

arkarsg committed:
Fix formatting issues with DukeIO

arkarsg committed:
Repackage classes into `duke` package

arkarsg committed:
Change processCommands from public to private

arkarsg committed:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

* commit '556af3f47a96b32898ab4cdbd65b16486a4871e8':
  Add Gradle support

arkarsg committed:
Clean up directories

9fc70c892 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,60 @@
+            taskList.loadFrom(storage.load());
+        } catch (FileNotFoundException p) {
+            dukeIo.echoMessage("Nothing to load");
+        }
```
What would happen if the file was present but contains invalid items? (partially or fully)

9fc70c892 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,60 @@
+        processCommands();        
+    }
+
+    private static void processCommands() 
```
I love how you have done up this portion

9fc70c892 commented on src/main/java/duke/InvalidCommandException.java:
```diff
@@ -0,0 +1,22 @@
+
+public class InvalidCommandException extends Exception {
+
+    private static final String HELP = " View all tasks with `list`";
+
+    public static final String TASK_NOT_FOUND_EXCEPTION = "[ERROR] : No such task." + HELP;
+    public static final String MARK_FORMAT_EXCEPTION = "[ERROR] : Mark/unmark tasks with task number." 
+                                                        + HELP;
+    public static final String NAME_FORMAT_EXCEPTION = "[ERROR] : Missing task name. ";
+    public static final String ARG_FORMAT_EXCEPTION = "[ERROR] Please specify date and time. ";
+    public static final String NULL_DELETE_EXCEPTION = "[ERROR] Missing task name to delete. Please specify a task."
+                                                        + HELP;
+    public static final String DELETE_FORMAT_EXCEPTION = "[ERROR] Delete task with task number."
+                                                        + HELP;
```
I love the use of Static final strings.

9fc70c892 commented on src/main/java/duke/MyDuke.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    private void populateCommands() {
+        cmdMap.put("list", (tokens) -> dukeIo.showAll());
+        cmdMap.put("todo", (tokens) -> addTodo(tokens));
+        cmdMap.put("deadline", (tokens) -> addDeadline(tokens));
+        cmdMap.put("event", (tokens) -> addEvent(tokens));
+        cmdMap.put("mark", (tokens) -> toggle(tokens));
+        cmdMap.put("unmark", (tokens) -> toggle(tokens));
+        cmdMap.put("delete", (tokens) -> delete(tokens));
```
I like the use of consumers here

9fc70c892 commented on src/main/java/duke/MyDuke.java:
```diff
@@ -0,0 +1,198 @@
+    public void exec(String[] tokens, TaskList taskList) throws InvalidCommandException {
+        allTasks = taskList;
+        try {
+            cmdMap.get(tokens[0]).accept(tokens);
```
What would happen if a user types
```
tOdO homework
```

9fc70c892 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,60 @@
+                dukeIo.showPrompt();
+            }
+        }
+        storage.saveFrom(taskList.getAllTasks());
```
What if an Exception was thrown here?

9fc70c892 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+
+    public Deadline(String desc, String deadline) {
+        super(desc);
+        setDeadline(deadline);
```
I like how this part is done. You have inspired me to have my code perform in this manner instead of taking in a LocalDateTime.

9fc70c892 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+                                                        ofPattern("dd/MM/yyyy HHmm");
+    private static final DateTimeFormatter OUT_FORMAT = DateTimeFormatter.
+                                                        ofPattern("dd LLL, h:mma");
+
```
Perhaps JavaDoc would be good here

9fc70c892 commented on src/main/java/duke/DukeIO.java:
```diff
@@ -0,0 +1,130 @@
+    private static final String REPLY = "|    ";
+    
+    
+    private static final String HELLO = REPLY + "Hello from\n" 
+                                        + SEPERATOR
+                                        + LOGO
+                                        + SEPERATOR
+                                        + REPLY + "What's on your mind today?\n"
+                                        + SEPERATOR;
```
Very inspirational. I might suggest this to other students.

9fc70c892 commented on src/main/java/duke/DukeIO.java:
```diff
@@ -0,0 +1,130 @@
+        System.out.println(SAVE_SUCCESS);
+    }
+    
+}
```
Overall very good naming conventions. Good job!

9fc70c892 commented on src/main/java/duke/MyDuke.java:
```diff
@@ -0,0 +1,198 @@
+        int byIndex = t.indexOf("/by");
+
+        try {
+            if (byIndex == -1) {
+                throw new InvalidCommandException(
+                    InvalidCommandException.ARG_FORMAT_EXCEPTION
+                    + "\n" + Deadline.showFormat());
+            }
+            if (byIndex + 1 == t.size()) {
+                throw new InvalidCommandException(
+                    InvalidCommandException.ARG_FORMAT_EXCEPTION
+                    + "\n" + Deadline.showFormat());
+            }
+            if (byIndex == 1) {
+                throw new InvalidCommandException(
+                    InvalidCommandException.NAME_FORMAT_EXCEPTION
+                    + "\n" + Deadline.showFormat());
+            }
```
Good catch of various potential exceptions!

9fc70c892 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,44 @@
+
+    @SuppressWarnings("unchecked")
+    public ArrayList<Task> load() throws IOException, ClassNotFoundException {
+        fis = new FileInputStream("./data/duke.txt");
```
Not sure if this would affect this but, what would happen if the file did not exist?

9fc70c892 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,44 @@
+        taskCount = tasks.size();
+    }
+
+    public static TaskList ofNull() {
+        return new TaskList(new ArrayList<Task>());
+    }
```
Good add!

9fc70c892 reviewed (commented):
Hi, I am Ed, I would be doing code review of your PR today.

Wonderful and inspiring. I cannot notice any code style errors. Naming is very good.

Good luck for iP and tP!

arkarsg committed:
Tests TaskList and ToDo classes with JUnit testing

arkarsg committed:
Repacks Duke into more packages for better abstraction

arkarsg committed:
Fixes minor formatting

arkarsg committed:
Implements find, writes JavaDoc and follows coding standard.

arkarsg commented:
> Hi, I am Ed, I would be doing code review of your PR today.
> 
> Wonderful and inspiring. I cannot notice any code style errors. Naming is very good.
> 
> Good luck for iP and tP!

Hey Ed! Thank you for the meaningful review. Just for context, at the point of your review, I have yet to add JavaDocs. But your review helped to catch some bugs or point out where I should elaborate on my JavaDocs.

arkarsg committed:
Revert "Implements find, writes JavaDoc and follows coding standard."

This reverts commit fc1320f73b3e827f69e15be2ddff60f0db7023c9.

arkarsg committed:
Adds JavaDoc for duke.Tasks and duke.Duke

arkarsg committed:
Add JavaDoc to duke.utils package.

arkarsg committed:
Add JavaDoc for duke.InvalidCommandException

arkarsg committed:
Fixes formatting to follow Java Coding Standard

arkarsg committed:
Implements "find" function.

arkarsg committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fixes formatting to follow Java Coding Standard

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/utils/DukeIO.java

arkarsg committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Implements "find" function.

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/utils/DukeIO.java
#	src/main/java/duke/utils/MyDuke.java
#	src/main/java/duke/utils/Storage.java

arkarsg committed:
Resolve additional confilcts

arkarsg committed:
Fix DukeIo

arkarsg committed:
Fix DukeIo.java

arkarsg committed:
A-CheckStyle

arkarsg committed:
Adds InvalidCommandException

arkarsg committed:
Preliminary implementation of GUI for Duke

arkarsg committed:
Refactors Duke.java

arkarsg committed:
Completes Level-10. Refactors for code quality.

arkarsg committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Completes Level-10. Refactors for code quality.
  Refactors Duke.java
  Preliminary implementation of GUI for Duke

arkarsg committed:
Rename "Tasks" directory to "tasks".

The folder for package named "Tasks" does not follow Coding Standard and hence, renamed to "tasks".

arkarsg committed:
Fixes GUI issues.

GUI does not handle long outputs properly and wraps text. This leaves user with missing information.
Responses are now `Text` instead of `Label` to work around this issue.

`Text` as response is more suitable as `Label` are meant for controls. `Text` also allows dialog boxes to resize according to the length of the response.

arkarsg committed:
Implements a custom welcome message shown at start up.

arkarsg committed:
Uses `assert` to test at various points of the code.

Assert is used in Parser class to check the assumption that methods parses inputs correctly. This is done by checking that the number of arguments returned after parsing equals the expected number of arguments. Assert is also used in Command classes. As retrieving arguments is hardcoded with the expected index of the arguments, it should be asserted that there are only the expected arguments present.

arkarsg committed:
Updates JavaDoc and CodeQuality

Deviating from the coding standards, there are missing JavaDocs and code quality with room for improvement.

Let's
- Add JavaDoc and header comments for all files
- Refactor condition checks as its own methods

arkarsg committed:
Merge pull request #2 from arkarsg/branch-A-Assertions

Uses `assert` to test at various points of the code.

arkarsg committed:
Merge branch 'master' into branch-A-CodeQuality

arkarsg committed:
Merge pull request #3 from arkarsg/branch-A-CodeQuality

Updates JavaDoc and CodeQuality

arkarsg committed:
Updates JavaDoc

arkarsg committed:
Implements a simple `sort` command

Simple sort feature is implemented. Currently, it is designed to accept a `key` as an argument to sort the list of tasks. This mutates the original task list and therefore, affects the indexes of the next operations.

SortCommand constructor takes in `key` of type String. `Key` is used to return a `Comparator<Task> object` which is consumed by the TaskList to compare tasks in the task list.

arkarsg committed:
Adds ability to sort items under `C-Sort`

Duke is now able to sort tasks in the task list chronologically (from nearest upcoming task to the furthest upcoming task) and sort lexicographically by task description.

This is done by implementing a subpackage `taskcomparator` under tasks. `NameComparator` compares the description of the Task. `DateComparator` compares the dates of the tasks if Task has no date associated to it, Task is sorted lexicographically.

As a design choice, only the `start date` is used to compare between tasks as we are assuming that the user wishes to know which task will come up the next immediate.

arkarsg committed:
Minor bug fixes and formatting

arkarsg committed:
Adds auto-load message for user experience

arkarsg committed:
Updates README.md and adds Ui.png

arkarsg committed:
Adds config file for formatting consistency

arkarsg committed:
Fixes README.md formatting issues

arkarsg committed:
Update _config.yml

arkarsg committed:
Update README.md

arkarsg committed:
Merge commit '5b6b4f44a9edc7b9e2f2cde7899fa97481aeccce'

* commit '5b6b4f44a9edc7b9e2f2cde7899fa97481aeccce':
  Update _config.yml

arkarsg committed:
Update README.md

arkarsg committed:
Update README.md

arkarsg committed:
Update README.md

arkarsg committed:
Update README.md

arkarsg committed:
Fixes minor coding standard
